File: src/main/java/org/team2679/frc2019/subsystems/Drive.java
package org.team2679.frc2019.subsystems;

import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
import edu.wpi.first.wpilibj.SpeedControllerGroup;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import org.team2679.framework.loops.Loop;
import org.team2679.framework.loops.Looper;
import org.team2679.framework.subsystems.Subsystem;
import org.team2679.frc2019.ControlPanel;
import org.team2679.logging.Logger;

public enum Drive implements Subsystem {

    INSTANCE;

    private WPI_TalonSRX FrontLeftMotor = new WPI_TalonSRX(1);
    private WPI_TalonSRX FrontRightMotor = new WPI_TalonSRX(4);
    private WPI_TalonSRX RearLeftMotor = new WPI_TalonSRX(3);
    private WPI_TalonSRX RearRightMotor = new WPI_TalonSRX(2);

    private SpeedControllerGroup leftGroup = new SpeedControllerGroup(FrontLeftMotor, RearLeftMotor);
    private SpeedControllerGroup rightGroup = new SpeedControllerGroup(FrontRightMotor, RearRightMotor);

    private DifferentialDrive driveTrain = new DifferentialDrive(leftGroup, rightGroup);

    private STATE currentState = null;

    public enum STATE{
        OPEN_LOOP, STOP, REVERSE
    }

    private Loop loop = new Loop() {
        @Override
        public void onStart(double time) {
            setState(STATE.STOP);
        }

        @Override
        public void onLoop(double time) {
            if(currentState == STATE.OPEN_LOOP){
                driveTrain.arcadeDrive(-ControlPanel.joystick.getAxis(1), ControlPanel.joystick.getAxis(0));
            }
            else if(currentState == STATE.STOP){
                driveTrain.arcadeDrive(0,0);
            }
        }

        @Override
        public void onStop(double time) {

        }
    };

    public void registerLoop(Looper looper){
        looper.register(this.loop);
    }

    public void setState(STATE state){
        if(currentState != state){
            this.currentState = state;
            Logger.INSTANCE.logINFO("state switched to " + state.toString(), "robot", "drive");
        }
    }

    @Override
    public void outputToSmartDashboard() {

    }

    @Override
    public void stop() {

    }

    @Override
    public void zeroSensors() {

    }
}

File: robot/src/main/kotlin/org/team2679/frc2019/Subsystems/Drive.kt
package org.team2679.frc2019.Subsystems

import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX
import edu.wpi.first.wpilibj.SpeedControllerGroup
import edu.wpi.first.wpilibj.drive.DifferentialDrive
import org.team2679.dashboard.Dashboard
import org.team2679.framework.states.Machine
import org.team2679.framework.states.machine
import org.team2679.framework.subsystems.Subsystem
import org.team2679.frc2019.ControlPanel
import org.team2679.logging.Logger

object Drive : Subsystem() {

    enum class States {
        OpenLoop, ReverseOpenLoop, KidMode, Auto, Stop
    }

    init{
        Logger.INSTANCE.logINFO("started drive subsystem", "robot", "subsystems")
    }

    private val FrontLeftMotor: WPI_TalonSRX = WPI_TalonSRX(0)
    private val FrontRightMotor: WPI_TalonSRX = WPI_TalonSRX(1)
    private val RearLeftMotor: WPI_TalonSRX = WPI_TalonSRX(2)
    private val RearRightMotor: WPI_TalonSRX = WPI_TalonSRX(3)

    private val leftGroup: SpeedControllerGroup = SpeedControllerGroup(FrontLeftMotor, RearLeftMotor)
    private val rightGroup: SpeedControllerGroup = SpeedControllerGroup(FrontRightMotor, RearRightMotor)

    private val driveTrain: DifferentialDrive = DifferentialDrive(leftGroup, rightGroup)

    override var machine: Machine = machine(States.OpenLoop) {
        whenState(States.OpenLoop) {
            execute {
                Logger.INSTANCE.logINFO("Open loop state started", "robot", "subsystem","drive")
            }
            repeat {
                outputToSmartDashboard()
                driveTrain.arcadeDrive(-ControlPanel.joystick.getAxis(1), ControlPanel.joystick.getAxis(0))
            }
        }

        whenState(States.ReverseOpenLoop) {
            execute {
                Logger.INSTANCE.logINFO("Reverse open loop state started", "robot", "subsystem","drive")
            }
            repeat {
                outputToSmartDashboard()
                driveTrain.arcadeDrive(ControlPanel.joystick.getAxis(1), ControlPanel.joystick.getAxis(0))
            }
        }
        whenState(States.Stop) {
            execute {
                Logger.INSTANCE.logINFO("stop state started", "robot", "subsystem","drive")
            }
            repeat {
                outputToSmartDashboard()
                driveTrain.arcadeDrive(0.0, 0.0)
            }
        }
        whenState(States.KidMode) {
            execute {
                Logger.INSTANCE.logINFO("kidmode state started", "robot", "subsystem","drive")
            }
            repeat {
                outputToSmartDashboard()
                driveTrain.arcadeDrive(-ControlPanel.joystick.getAxis(1)*0.5, ControlPanel.joystick.getAxis(0)*0.5)
            }
        }
    }

    fun setState(state: States) {
        TODO("add setState to the machine object and implement this method")
    }

    override fun outputToSmartDashboard() {
        Dashboard.INSTANCE.putDouble("Drive X Value", ControlPanel.joystick.getAxis(0));
        Dashboard.INSTANCE.putDouble("Drive Y Value", ControlPanel.joystick.getAxis(1));
    }

    override fun zeroSensors() {
    }

    override fun stop() { setState(States.Stop) }

    fun doDriveMode() { setState(States.OpenLoop) }

    fun doReverseMode() { setState(States.ReverseOpenLoop) }
}